src/callbacks/kfac/kfac.cpp

```diff
@@ -0,0 +1,482 @@
+      prof_region_begin(("kfac-update/average/" + block->get_name()).c_str(), prof_color, prof_sync);
+      block->update_kronecker_average(
+          comm,
+          m_kronecker_decay,
+          m_print_matrix, m_print_matrix_summary);
+      prof_region_end(("kfac-update/average/" + block->get_name()).c_str(), prof_sync);
+    }
+    prof_region_end("kfac-update/average", prof_sync);
+
+    prof_region_end("kfac-update", prof_sync);
+  }
+
+  // Step 2: Model-parallel inverse computation
+  prof_region_begin("kfac-inverse", prof_color, prof_sync);
+  for(auto& block : m_blocks) {
+    if(!is_kronecker_update_required || (size_t) comm->get_rank_in_trainer() != block->get_inverse_proc_rank())
+      continue;
+
+    prof_region_begin(("kfac-inverse/" + block->get_name()).c_str(), prof_color, prof_sync);
+    // TODO: Add kfac_block::is_bn?
+    const bool is_bn = dynamic_cast<kfac_block_bn<Device>*>(block.get()) != nullptr;
+    block->update_kronecker_inverse(
+        comm, m_use_pi,
+        is_bn ? m_damping_bn_act : m_damping_act,
+        is_bn ? m_damping_bn_err : m_damping_err,
+        m_learning_rate_factor,
+        m_print_matrix, m_print_matrix_summary,
+        m_print_time);
+    prof_region_end(("kfac-inverse/" + block->get_name()).c_str(), prof_sync);
+  }
+  m_has_kronecker_inverse = true;
+  prof_region_end("kfac-inverse", prof_sync);
+
+#ifdef LBANN_NVPROF
+  prof_region_begin("kfac-inverse-barrier", prof_color, prof_sync);
+  CHECK_CUDA(cudaDeviceSynchronize());
+  comm->trainer_barrier();
+  prof_region_end("kfac-inverse-barrier", prof_sync);
+#endif // LBANN_NVPROF
+
+  // Step 3: All-gather of each preconditioned gradient tensor
+  prof_region_begin("kfac-allgather", prof_color, prof_sync);
+  {
+    // List-up buffers to synchronize.
+    std::vector<std::pair<size_t, El::AbstractMatrix<DataType>*>> buffers;
+    int local_buffer_size = 0, global_buffer_size = 0;
+    for(auto& block : m_blocks)
+      for(auto L : block->get_preconditioned_grad_buffers()) {
+        const size_t rank = block->get_inverse_proc_rank();
+        buffers.emplace_back(rank, L);
+        assert(L->Width() == 1);
+        if(rank == (size_t) comm->get_rank_in_trainer())
+          local_buffer_size += L->Height();
+        global_buffer_size += L->Height();
+      }
+
+    // Perform allgather.
+    const auto allgather_mode = kfac_allgather_mode::ALLREDUCE;
+    const auto is_buffer_needed = kfac_util::is_allgather_buffer_required(allgather_mode);
+    El::Matrix<DataType, Device>& local_buffer =
+        get_workspace_matrix(
+            "allgather_send_buffer",
+            is_buffer_needed.first ? local_buffer_size : 0,
+            1);
+    El::Matrix<DataType, Device>& global_buffer =
+        get_workspace_matrix(
+            "allgather_recv_buffer",
+            is_buffer_needed.second ? global_buffer_size : 0,
+            1);
+    kfac_util::allgather_blocks(
+        buffers, local_buffer, global_buffer, comm, allgather_mode);
```

Comment:
We are preconditioning the gradient at the same time and on the same process as we are inverting the Fisher matrix, so it makes sense to allgather the preconditioned gradients to everyone. However, if we want to precondition the gradient at every step (using a stale Fisher matrix), would it be better to perform the preconditioning in parallel? In other words, all processes should have a copy of all the Fisher matrix inverses so that they can apply them to the gradient without needing any communication. The existing scheme is better if the preconditioning is expensive, the Fisher matrices are large, and the gradients are small. Were these your considerations?

Reply:
One of the advantages of distributing Fisher matrix computation among processes is that it can distribute both the computational workload and the memory consumption. For large dense DNNs, such as the ATOM model, each Fisher matrix inverse consumes memory of the same order of magnitude as the number of parameters, which may take few GBs. So I think it is a good idea to optionally remove all-gather for smaller networks, but I think this function is still possible for some large networks that take a large amount of memory.